exchanges/okx/okx_websocket.go

```diff
@@ -337,16 +332,22 @@ func (ok *Okx) WsAuth(ctx context.Context, dialer *websocket.Dialer) error {
 	}
 }
 
-// wsFunnelConnectionData receives data from multiple connection and pass the data
-// to wsRead through a channel responseStream
-func (ok *Okx) wsFunnelConnectionData(ws stream.Connection) {
+// wsReadData sends msgs from public and auth websockets to data handler
+func (ok *Okx) wsReadData(ws stream.Connection) {
 	defer ok.Websocket.Wg.Done()
 	for {
 		resp := ws.ReadMessage()
 		if resp.Raw == nil {
 			return
 		}
-		responseStream <- stream.Response{Raw: resp.Raw}
+		// Ensure that websocket shutdown will block until WsHandleData returns, so that the multiplexer is safe from new blocking channel sends.
+		ok.Websocket.Wg.Add(1)
+		go func(msg []byte) {
+			defer ok.Websocket.Wg.Done()
+			if err := ok.WsHandleData(msg); err != nil {
+				ok.Websocket.DataHandler <- err
+			}
+		}(resp.Raw)
```

Comment:
Now to go further, I wrote the following:
```go
var indexPerMessage, indexPerGoRoutine int64

// wsReadData sends msgs from public and auth websockets to data handler
func (ok *Okx) wsReadData(ws stream.Connection) {
	defer ok.Websocket.Wg.Done()
	for {
		resp := ws.ReadMessage()
		if resp.Raw == nil {
			return
		}
		indexPerMessage++
		localIndexPerMessage := indexPerMessage
		// Ensure that websocket shutdown will block until WsHandleData returns, so that the multiplexer is safe from new blocking channel sends.
		ok.Websocket.Wg.Add(1)
		go func(msg []byte, passedInIndex int64) {
			indexPerGoRoutine++
			defer ok.Websocket.Wg.Done()
			if err := ok.WsHandleData(msg); err != nil {
				ok.Websocket.DataHandler <- err
			}
			if passedInIndex != indexPerGoRoutine {
				log.Warnf(log.ExchangeSys, "out of order index indexPerGoRoutine %v passedInIndex %v", indexPerGoRoutine, passedInIndex)
			}
		}(resp.Raw, localIndexPerMessage)
	}
}

```
Maybe I've written a bad test. This is without an authconnection. What I'm doing here is that `ReadMessage` is sequential, so the `indexPerMessage` denotes the latest update. I localise it to prevent anything weird, then pass it by value to the goroutine. In theory, these should always match if its FIFO right? But they don't when there is sufficient load

Further, I also put a `panic` when an orderbook checksum failed validation and it was able to be hit only when there was the goroutine. So we can't process updates out of order

Reply:
Okay. I'll institute a FIFO.
I would never be worried about blowing up the number of goros.
I've frequently run with hundreds of thousands with no issue at all. It's literally one of the cornerstones of golang.
In this case they're not even long running, so I wouldn't hesitate to do this at all for that reason.

The reason I don't want to FIFO everything is because we don't want to block processing incoming order txs because of resource contention for orderbook and ticker.
Like you say, okx can be busy. That's the exact reason I don't want to put everything in the same fifo.

I think I'll look for a compromise downstream.

Note: Pretty sure we can't/shouldn't wg.Add before reading, because it's a blocking call, right?

And one final note: I get okx checksum errors constantly right now:
```
2023-06-30T12:26:43.169+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
2023-06-30T12:26:44.311+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
2023-06-30T12:26:51.119+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
2023-06-30T12:26:57.639+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
```
It's on my list to fix. They don't usually start straight away but get worse after a few hours.
Maybe I'll role that into this fix if at all possible. I suspect it's still related to dropped packets, despite a very good connection here.

Reply:
I've confirmed you contention and issue.

I've switched to blocking processing for this PR.

I'll work on breaking out the messages which take longer another day.
It's mostly orderbook which can block for up to 4ms. We definitely can't have that.